parsedtree.cpp

```diff
@@ -0,0 +1,154 @@
+/****************************************************************************
+**
+** Copyright (C) 2017 N7 Mobile sp. z o. o.
+** Contact: http://n7mobile.com/Space
+**
+** This file is part of ASN.1/ACN Plugin for QtCreator.
+**
+** Plugin was developed under a programme and funded by
+** European Space Agency.
+**
+** This Plugin is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** This Plugin is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with this program.  If not, see <http://www.gnu.org/licenses/>.
+**
+****************************************************************************/
+
+#include "parsedtree.h"
+
+#include "utils/qtcassert.h"
+
+#include "parseddatastorage.h"
+
+using namespace Asn1Acn::Internal;
+
+static ParsedTree *m_instance;
+static QMutex m_instanceMutex;
+
+ParsedTree *ParsedTree::instance()
+{
+    if (m_instance)
+        return m_instance;
+
+    QMutexLocker locker(&m_instanceMutex);
+
+    if (!m_instance)
+        m_instance = new ParsedTree;
+
+    return m_instance;
+}
+
+ParsedTree::ParsedTree()
+{
+    m_treeRoot = ParsedTreeNode::ParsedTreeNodePtr(new ParsedTreeNode);
+
+    ParsedDataStorage *storage = ParsedDataStorage::instance();
+    connect(storage, &ParsedDataStorage::fileUpdated,
+            this, &ParsedTree::updateParsedTreeNode);
+}
+
+void ParsedTree::addProjectNode(ParsedTreeNode::ParsedTreeNodePtr projectNode)
+{
+    QMutexLocker locker(&m_dataMutex);
+
+    m_treeRoot->addChild(projectNode);
+
+    locker.unlock();
+
+    emit treeUpdated();
```

Comment:
what do you think about such notation (instead of explicit `unlock` call):
```
{
 QMutexLocker locker(&m_dataMutex);
 m_treeRoot->addChild(projectNode);
}
emit treeUpdated();
```